frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+
+			if T::TargetList::contains(&nomination) {
+				let _ = T::TargetList::on_update(&nomination, new_stake).defensive();
+			}
+		};
+
+		for nomination in new {
+			// Create a new entry if it does not exist
+			let new_stake = Self::approval_stake(&nomination)
+				.defensive_ok_or(Error::<T>::DoesNotExist)?
+				.saturating_add(Self::slashable_balance_of(who));
+
+			update_approval_stake(&nomination, new_stake);
+		}
+
+		for nomination in obsolete {
+			// we should actually fail if an old nomination was not in the map, which should never
+			// happen.
+			let new_stake = Self::approval_stake(&nomination)
+				.defensive_ok_or(Error::<T>::DoesNotExist)?
+				.saturating_sub(Self::slashable_balance_of(who));
+
+			update_approval_stake(&nomination, new_stake);
+		}
+		let _ =
+			T::VoterList::on_insert(who.clone(), Self::to_vote(Self::slashable_balance_of(who)))
+				.defensive_unwrap_or_default();
+		Ok(())
+	}
+
+	/// This should only be called if that stash isn't already a validator. Note, that if we want to
+	/// properly track ApprovalStake here - we need to make sure we subtract the validator stash
+	/// balance when they chill?
+	///	Why? Because we don't remove ApprovalStake when a validator chills and we need to make sure
+	/// their self-stake is up-to-date and not applied twice.
+	fn on_validator_add(who: &T::AccountId) -> DispatchResult {
+		let self_stake = Self::slashable_balance_of(who);
+		let new_stake = Self::approval_stake(who).unwrap_or_default().saturating_add(self_stake);
+
+		// maybe update sorted list.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(self_stake)).defensive();
+
+		// TODO: Make sure this always works. Among other things we need to make sure that when the
+		// migration is run we only have active validators in TargetList and not the chilled ones.
+		let _ = T::TargetList::on_insert(who.clone(), new_stake).defensive();
+		ApprovalStake::<T>::set(who, Some(new_stake));
+
+		Ok(())
+	}
+
+	fn on_validator_remove(who: &T::AccountId) -> DispatchResult {
+		let _ = T::TargetList::on_remove(who).defensive();
+		let _ = T::VoterList::on_remove(who).defensive();
+
+		ApprovalStake::<T>::mutate(who, |x: &mut Option<BalanceOf<T>>| {
+			x.map(|b| b.saturating_sub(Self::slashable_balance_of(who)))
+		})
+		.ok_or(Error::<T>::DoesNotExist)?;
+
+		Ok(())
+	}
+
+	fn on_nominator_remove(who: &T::AccountId, nominations: Vec<T::AccountId>) -> DispatchResult {
+		let score = Self::slashable_balance_of(who);
+		let _ = T::VoterList::on_remove(who).defensive();
+		for validator in nominations {
+			ApprovalStake::<T>::mutate(&validator, |x: &mut Option<BalanceOf<T>>| {
+				x.map(|b| b.saturating_sub(score))
+			})
+			.ok_or(Error::<T>::DoesNotExist)
+			.defensive()?;
+		}
+
+		Ok(())
+	}
+
+	fn on_reaped(who: &T::AccountId) -> DispatchResult {
+		ApprovalStake::<T>::remove(who);
+
+		Ok(())
+	}
+}
+
+impl<T: Config> ScoreProvider<T::AccountId> for Pallet<T> {
```

Comment:
This pallet is importantly missing an implementation -- nay, in fact, two implementations of `SortedListProvider`. Then the full end to end picture, as depicted here, would come into perspective.

Reply:
I have opted for a simpler solution that has to do with ROSortedListProvider. We don't need to implement any of this anymore, instead we use bags-list instances that we pass to both tracker and staking.